### PR TITLE
Add attributes to disable L3 rewrites

### PR DESCRIPTION
**Pull request by vivekmoorthy**
Knobs for disabling rewrites to following fields
- Src MAC disable
- Dst MAC disable
- Vlan rewrite disable

While we do L3 IP based routing, we have scenarios where we need knobs for disabling L2 field rewrites.

Case 1: For some scenarios, the switch serves as an L3 passthrough. Creating nexthop/RIF for every source mac, dst-mac, vlan will not scale, and we would like to simply passthrough the L2 fields without any rewrites
Case 2: For some scenarios, we set VLAN in pre-ingress stage based on certain classification, and would like to preserve this VLAN in the packet.

### CODE REVIEW

**Comment by kcudnik** [comment]
please fix build errors

**Comment by vivekmoorthy** [comment]
> please fix build errors

Done.

**Comment by vivekmoorthy** [comment]
> please fix build errors

done

**Comment by vivekmoorthy** [comment]
@kcudnik @mikeberesford could you review this PR

**Comment by mikeberesford** [comment]
@vivekmoorthy we will need to bring this to the SAI community meeting  next week. I can help coordinate

File: inc/sainexthop.h [hunk 1]
```diff
@@ -226,6 +226,33 @@ typedef enum _sai_next_hop_attr_t
...
     SAI_NEXT_HOP_ATTR_QOS_TC_AND_COLOR_TO_MPLS_EXP_MAP,
 
+    /**
+     * @brief To enable/disable source MAC rewrite
+     *
+     * @type bool
+     * @flags CREATE_AND_SET
+     * @default false
+     */
+    SAI_NEXT_HOP_ATTR_DISABLE_SRC_MAC_REWRITE,
```
Comment (srikrishnagopu) [inline inc/sainexthop.h:236]: Couple of questions here:
-  Is the idea here to perform L3 routing but not rewrite the packet ?
- Would it be possible to update the PR description with the usecase ?

**Review by kcudnik — APPROVED** [review]

File: inc/sainexthop.h (hunk 1 above)
Comment (vivekmoorthy) [reply inc/sainexthop.h:236]: yes we rely on L3 based IP forwarding, and have some scenarios for passing through without rewrites. I have updated the PR description.

**Review by srikrishnagopu — APPROVED** [review]

**Comment by rlhui** [comment]
Today in SAI pipeline for L3/routed packets, dst mac/src mac/vlan id are replaced in neighbor encap/egress rif block.  We discussed in SAI Meeting if these knobs should be in neighbor or rif object instead.  However, if disable L3 rewrite are only for certain IP flows and not for entire egress IP interface/neighbor, it'd still be better to stay at next hop object.

**Review by srikrishnagopu — CHANGES_REQUESTED** [review]
Moving it to request based on today's feedback.

**Comment by vivekmoorthy** [comment]
> Moving it to request based on today's feedback.

Updated link to document with requirements, proposal and example in description.

**Comment by vivekmoorthy** [comment]
@JaiOCP @rlhui @srikrishnagopu @rck-innovium Pls review. I have added link document in description with example usage.

**Review by JaiOCP — APPROVED** [review]
Looks good to me

**Review by ashutosh-agrawal — APPROVED** [review]
LGTM

**Comment by rck-innovium** [comment]
@vivekmoorthy  Please check in the document as MD file.

File: inc/sainexthop.h [hunk 2]
```diff
@@ -226,6 +226,33 @@ typedef enum _sai_next_hop_attr_t
...
+    SAI_NEXT_HOP_ATTR_DISABLE_DST_MAC_REWRITE,
+
+    /**
+     * @brief To enable/disable VLAN rewrite
+     *
+     * @type bool
+     * @flags CREATE_AND_SET
+     * @default false
+     */
+    SAI_NEXT_HOP_ATTR_DISABLE_VLAN_REWRITE,
```
Comment (marian-pritsak) [inline inc/sainexthop.h:254]: What if the result of routing lookup is router port, not a member of the VLAN?

**Review by marian-pritsak — CHANGES_REQUESTED** [review]

**Review by srikrishnagopu — APPROVED** [review]
LGTM.

**Comment by vivekmoorthy** [comment]
> @vivekmoorthy  Please check in the document as MD file.

Done.

**Comment by vivekmoorthy** [comment]
@rck-innovium @marian-pritsak could you pls review

**Comment by vivekmoorthy** [comment]
@rlhui Based on discussion last week and with the approvals and comments addressed, could we merge this PR?

**Comment by rlhui** [comment]
@marian-pritsak , please help review/approve? Thanks.

**Comment by vivekmoorthy** [comment]
> @marian-pritsak , please help review/approve? Thanks.

@marian-pritsak could you please help with this?
Thanks.

**Pull request merged**